src/ucp/core/ucp_ep.c

```diff
@@ -807,32 +807,79 @@ void ucp_ep_disconnected(ucp_ep_h ep, int force)
 unsigned ucp_ep_local_disconnect_progress(void *arg)
 {
     ucp_request_t *req         = arg;
-    ucp_ep_h ep                = req->send.ep;
-    ucs_async_context_t *async = &ep->worker->async; /* ep becomes invalid */
+    ucs_async_context_t *async = &req->send.ep->worker->async;
 
     ucs_assert(!(req->flags & UCP_REQUEST_FLAG_COMPLETED));
 
     UCS_ASYNC_BLOCK(async);
-    ucs_debug("ep %p: disconnected with request %p, %s", ep, req,
-              ucs_status_string(req->status));
-    ucp_ep_disconnected(ep, req->send.flush.uct_flags & UCT_FLUSH_FLAG_CANCEL);
+    if (req->send.ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID) {
```

Comment:
how could it be that close req is not valid? it's being used here already to find the ep?
also, better not call callback inside the async lock

Reply:
> how could it be that close req is not valid?

in else branch, this is flush requested returned from ep_close/internal_flush

 it's being used here already to find the ep?

yes, will restore old code

> also, better not call callback inside the async lock

👍 in addition need to schedule request completion from `ucp_ep_close_request_complete` to main thread since it can be invoked from async handler by timer